### PR TITLE
[4.0] Fixed workflow conditions in search plugin

### PR DESCRIPTION
**Pull request by bahl24**
Pull Request for Issue #24479  .

### Summary of Changes
No results due to incorrect SQL query.

### Expected result
Search page with results which are published

### Actual result
Search always returns 0 results

### CODE REVIEW

**Comment by brianteeman** [comment]
ws.condition=0 is the UNPUBLISHED state so that change cannot be correct can it?

**Comment by bahl24** [comment]
I tested and found that ws.condition=0 to be working.

**Comment by brianteeman** [comment]
That means there is a different problem to fix

**Comment by bahl24** [comment]
I think I got the required solution.
Joining by ID always resulted in UNPUBLISHED state
Screenshot from 2019-04-04 22-16-55

**Comment by brianteeman** [comment]
@bembelimen can you check this please

**Comment by bembelimen** [comment]
To get the correct state, you should always join over ```#__workflow_associations```

**Comment by bahl24** [comment]
Updated: `wa.stage_id=2` is equivalent to `ws.condition=2` as both refer to PUBLISHED state

**Comment by bembelimen** [comment]
> 
> 
> Updated: `wa.stage_id=2` is equivalent to `ws.condition=2` as both refer to PUBLISHED state

That's only for the default workflow the case (and only by chance), see e.g. for the correct join: https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_content/Model/ArticlesModel.php#L231-L260

**Comment by brianteeman** [comment]
> Updated: wa.stage_id=2 is equivalent to ws.condition=2 as both refer to PUBLISHED state

Only in the default workflow!!!

**Comment by bembelimen** [comment]
You shouldn't use the magic number "1" but the extension class constant for conditions.

**Comment by bahl24** [comment]
> You shouldn't use the magic number "1" but the extension class constant for conditions.

Sorry, I didn't get you, `ws.condition=1` points to published state. Is there an example regarding the extension class constant method?

**Comment by Bakual** [comment]
The join on the #__workflow_associations table also needs to filter by `extension = "com_content"`. Otherwise you also join on items from other extensions.

**Comment by bembelimen** [comment]
See: https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_content/Model/ArticlesModel.php#L363-L364

and: https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_content/Model/ArticlesModel.php#L370

File: plugins/search/content/content.php [hunk 1]
```diff
@@ -292,10 +293,12 @@ public function onContentSearch($text, $phrase = '', $ordering = '', $areas = nu
 				->join('LEFT', '#__workflow_associations AS wa ON wa.item_id = a.id')
 				->join('INNER', '#__workflow_stages AS ws ON ws.id = wa.stage_id')
 				->where(
-					'(' . $where . ') AND ws.condition=1 AND c.published = 1 AND a.access IN (' . $groups . ') '
+					'(' . $where . ') AND c.published = 1 AND a.access IN (' . $groups . ') '
+						. 'AND (ws.condition = ' . ContentComponent::CONDITION_PUBLISHED .')'
 						. 'AND c.access IN (' . $groups . ')'
 						. 'AND (a.publish_up = ' . $db->quote($nullDate) . ' OR a.publish_up <= ' . $db->quote($now) . ') '
 						. 'AND (a.publish_down = ' . $db->quote($nullDate) . ' OR a.publish_down >= ' . $db->quote($now) . ')'
+						. 'AND (' . $db->quoteName("wa.extension") . '=' . $db->quote("com_content") . ')'
```
Comment (Bakual) [inline plugins/search/content/content.php]: Can you put that into an own ->where() clause? Like done a few lines above:
https://github.com/joomla/joomla-cms/blob/1980cef6ec7d57a433f58f6766f7cf8971b0ad92/plugins/search/content/content.php#L208-L211

I honestly don't know why it was done like this here, but we shouldn't expand that bad practice further 😄

Comment (Bakual) [reply plugins/search/content/content.php]: To be clear, just do like this:
````
->join('INNER', '#__workflow_stages AS ws ON ws.id = wa.stage_id')
->where($db->quoteName("wa.extension") . '=' . $db->quote("com_content"))
````

Comment (bahl24) [reply plugins/search/content/content.php]: @Bakual Changed, kindly check.

Comment (Bakual) [reply plugins/search/content/content.php]: Looks fine to me, thanks.

File: plugins/search/content/content.php [hunk 2]
```diff
@@ -15,6 +15,7 @@
 use Joomla\CMS\Plugin\CMSPlugin;
 use Joomla\CMS\Router\Route;
 use Joomla\Component\Search\Administrator\Helper\SearchHelper;
+use Joomla\Component\Content\Administrator\Extension\ContentComponent;
```
Comment (ReLater) [inline plugins/search/content/content.php:17]: Please sort `use` lines alphabetically. Line 18 before line 17 in this case.

Comment (bahl24) [reply plugins/search/content/content.php:17]: Done

File: plugins/search/content/content.php [hunk 3]
```diff
@@ -289,9 +290,12 @@ public function onContentSearch($text, $phrase = '', $ordering = '', $areas = nu
...
 				->innerJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
-				->join('LEFT', '#__workflow_stages AS ws ON ws.id = state')
+				->join('LEFT', '#__workflow_associations AS wa ON wa.item_id = a.id')
+				->join('INNER', '#__workflow_stages AS ws ON ws.id = wa.stage_id')
+				->where($db->quoteName('wa.extension') . '=' . $db->quote('com_content'))
 				->where(
-					'(' . $where . ') AND ws.condition=1 AND c.published = 1 AND a.access IN (' . $groups . ') '
+					'(' . $where . ') AND c.published = 1 AND a.access IN (' . $groups . ') '
+						. 'AND (ws.condition = ' . ContentComponent::CONDITION_PUBLISHED . ') '
 						. 'AND c.access IN (' . $groups . ')'
```
Comment (Quy) [inline plugins/search/content/content.php]:
```suggestion
						. 'AND c.access IN (' . $groups . ') '
```

**Comment by bahl24** [comment]
@Quy , thanks for pointing out. I've made the changes

File: plugins/search/content/content.php [hunk 4]
```diff
@@ -289,10 +290,13 @@ public function onContentSearch($text, $phrase = '', $ordering = '', $areas = nu
 				->select($db->quote('2') . ' AS browsernav')
 				->from($db->quoteName('#__content', 'a'))
 				->innerJoin($db->quoteName('#__categories', 'c') . ' ON c.id = a.catid')
-				->join('LEFT', '#__workflow_stages AS ws ON ws.id = state')
+				->join('LEFT', '#__workflow_associations AS wa ON wa.item_id = a.id')
+				->join('INNER', '#__workflow_stages AS ws ON ws.id = wa.stage_id')
+				->where($db->quoteName('wa.extension') . '=' . $db->quote('com_content'))
```
Comment (wilsonge) [inline plugins/search/content/content.php]: This needs to be in the join. else the inner join will give multiple items back if you have the same item id for more than one component i think

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**